### PR TITLE
Changes in accident_count_by_car_type #2071

### PR DESCRIPTION
**Pull request by BarVolunteering**
*Change table to VehicleMarkerView
*Order the data by segment
*Return only top 3

### CODE REVIEW

File: anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py [hunk 1]
```diff
@@ -22,8 +26,10 @@ def __init__(self, request_params: RequestParams):
         self.rank = 17
 
     def generate_items(self) -> None:
-        self.items = AccidentCountByCarTypeWidget.get_stats_accidents_by_car_type_with_national_data(
-            self.request_params
+        self.items = (
```
Comment (shaysw) [inline anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:29]: why this change? please revert if it's not needed. in general accidental moot changes make it more difficult to perform CRs

File: anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py [hunk 2]
```diff
@@ -46,57 +52,71 @@ def get_stats_accidents_by_car_type_with_national_data(
         data_by_segment = AccidentCountByCarTypeWidget.percentage_accidents_by_car_type(
             involved_by_vehicle_type_data
         )
-        national_data = AccidentCountByCarTypeWidget.percentage_accidents_by_car_type_national_data_cache(
-            start_time, end_time
+        national_data = (
```
Comment (shaysw) [inline anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:55]: same

File: anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py [hunk 3]
```diff
@@ -46,57 +52,71 @@ def get_stats_accidents_by_car_type_with_national_data(
...
+                        {"label_key": "percentage_country", "value": national_data[car_type]},
+                    ],
                 }
             )
         return out
 
     @staticmethod
     def percentage_accidents_by_car_type(involved_by_vehicle_type_data):
-        driver_types = defaultdict(float)
+        vehicle_type_dict = defaultdict(float)
```
Comment (shaysw) [inline anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:75]: why float?

File: anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py [hunk 4]
```diff
@@ -46,57 +52,71 @@ def get_stats_accidents_by_car_type_with_national_data(
...
             elif vehicle_type in VehicleCategory.BICYCLE_AND_SMALL_MOTOR.get_codes():
-                driver_types[VehicleCategory.BICYCLE_AND_SMALL_MOTOR.value] += count
+                vehicle_type_dict[VehicleCategory.BICYCLE_AND_SMALL_MOTOR.value] += count
             else:
-                driver_types[VehicleCategory.OTHER.value] += count
+                vehicle_type_dict[VehicleCategory.OTHER.value] += count
+
+        output = {}
+        # Order the vehicle_type_dict dict by value which is the count of the vehicle_type
+        vehicle_type_dict_ordered = dict(
```
Comment (shaysw) [inline anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:93]: you can perform the ordering on the original dict, no need for additional variable

File: anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py [hunk 5]
```diff
@@ -46,57 +52,71 @@ def get_stats_accidents_by_car_type_with_national_data(
...
-        for k, v in driver_types.items():  # Calculate percentage
-            output[k] = 100 * v / total_count
+        # Calculate percentage of the top 3
+        for i, (key, value) in enumerate(vehicle_type_dict_ordered.items()):
+            output[key] = 100 * value / total_count
+            if i >= TOP_3:
+                break
 
-        return output
+        # Use defaultdict to return 0 where there is no key, so when there is no data for this type of vehicle the
```
Comment (shaysw) [inline anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:103]: no need for newline

File: anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py [hunk 6]
```diff
@@ -46,57 +52,71 @@ def get_stats_accidents_by_car_type_with_national_data(
...
+        vehicle_type_dict_ordered = dict(
+            sorted(vehicle_type_dict.items(), key=lambda item: item[1], reverse=True)
+        )
 
-        output = defaultdict(float)
-        for k, v in driver_types.items():  # Calculate percentage
-            output[k] = 100 * v / total_count
+        # Calculate percentage of the top 3
+        for i, (key, value) in enumerate(vehicle_type_dict_ordered.items()):
+            output[key] = 100 * value / total_count
```
Comment (shaysw) [inline anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:99]: just my opinion: return percentages as int, not float

**Review by shaysw — APPROVED** [review]
Approved with suggestions/notes

File: anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py (hunk 3 above)
Comment (BarVolunteering) [reply anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:75]: That's the original interface, there was no request to change it

File: anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py (hunk 5 above)
Comment (BarVolunteering) [reply anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:103]: It's over the recommended line size:
image

If you want i can change it to a one line

**Review by atalyaalon — CHANGES_REQUESTED** [review]
@BarVolunteering looks good.
Please resolve conflicts.

File: translations/he/LC_MESSAGES/messages.po [hunk 7]
```diff
@@ -210,11 +210,11 @@ msgstr ""
...
 
-#: anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:114
+#: anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:134
 msgid ""
-"comparing vehicle type percentage in accidents in {} relative to national"
-" average"
-msgstr "השוואת אחוז הרכבים בתאונות במקטע {} לעומת ממוצע ארצי"
+"Comparing top 3 vehicle type percentage in accidents in {} relative to "
+"national average"
+msgstr "השוואת 3 המובילים באחוז הרכבים בתאונות במקטע {} לעומת ממוצע ארצי"
```
Comment (atalyaalon) [inline translations/he/LC_MESSAGES/messages.po:217]: Also - please verify title with product team - @MichalOren

Comment (MichalOren) [reply translations/he/LC_MESSAGES/messages.po:217]: Hi @BarVolunteering and @atalyaalon 
Does the query include all the severity of the accident. Including light, severe & fatal?
Why "percent" while the data appears as a numbers?
In  general....it should look like this
image

As for the filter for **top 3** results it does not appear in the title but in the "info" section
So  insert the following sentence:
התפלגות תאונות לפי סוגי רכב במקטע, בפרק הזמן הנבחר. מוצגים 3 סוגי הרכב הנפוצים ביותר בתאונות במקטע זה.

Comment (MichalOren) [reply translations/he/LC_MESSAGES/messages.po:217]: למעשה נראה שיש עוד הגדרות שחסרות. אבל אני לא רוצה לעכב את המשימה  שמוגדרת בתחילת הדף

Comment (BarVolunteering) [reply translations/he/LC_MESSAGES/messages.po:217]: The query include all severities of the accident.
We are using percent because thats the original interface before the change, we can change it if you want.

What do mean in 
"As for the filter for top 3 results it does not appear in the title but in the "info" section"? - the information section in the JSON is not in use, do you mean we should use it?

I will change the title

Comment (MichalOren) [reply translations/he/LC_MESSAGES/messages.po:217]: Hi Bar
- Keep current percentage values- no change.
- so the title probbly be
תאונות לפי סוג רכב  במקטע <> בהשוואה לכל הארץ
what do you think?

- kindly add the description above inside information section.  yes we should start using it in this widget
as i mentioned above this widget is missing some setting & definition

Comment (BarVolunteering) [reply translations/he/LC_MESSAGES/messages.po:217]: 1. בנוגע לכותרת "השוואת סוגי רכב המעורבים בתאונות במקטע <> בהשוואה לכל הארץ",  אין צורך לציין שאלו שלושת סוגי הרכבים עם הכי הרבה תאונות?
2. מה התוכן שאתם רוצים שיופיע בinformation ?
3. לא הבנתי את המשפט "this widget still missing some setting & definition", יש צורך להגדיר מחדש את המשימה?

**Review by atalyaalon — CHANGES_REQUESTED** [review]
looks good! I think the logic needs a small fix.

File: anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py [hunk 8]
```diff
@@ -33,10 +38,10 @@ def get_stats_accidents_by_car_type_with_national_data(
...
         if involved_by_vehicle_type_data is None:
             involved_by_vehicle_type_data = widget_utils.get_accidents_stats(
-                table_obj=InvolvedMarkerView,
-                filters=widget_utils.get_injured_filters(request_params.location_info),
-                group_by="involve_vehicle_type",
-                count="involve_vehicle_type",
+                table_obj=VehicleMarkerView,
+                filters=request_params.location_info,
+                group_by="vehicle_type",
+                count="vehicle_type",
```
[the remark this answers is not in the document]
Comment (ziv17) [reply anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:44]: Do we have same accident appearing multiple times from same provider?

[the remark this answers is not in the document]
Comment (BarVolunteering) [reply anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:44]: @atalyaalon I don't remember what provider_and_id mean, can you remind me please? 
In the past we had an excel with columns descriptions, do you remember where it is?
@ziv17 Don't know - @atalyaalon?

[the remark this answers is not in the document]
Comment (atalyaalon) [reply anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:44]: @BarVolunteering the spreadsheet [is here](https://docs.google.com/spreadsheets/d/1qaVV7NKXVYNmnxKZ4he2MKZDAjWPHiHfq-U5dcNZM5k/edit#gid=375223389)

provider_and_id is a unique identifier of an accident. There are cases where there is more than one vehicle of a certain type in an accident - but we want to count it only once, hence we want to count distinct provider_and_id.

File: anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py [hunk 9]
```diff
@@ -1,16 +1,21 @@
...
+from anyway.models import VehicleMarkerView
+from anyway.request_params import RequestParams
 from anyway.vehicle_type import VehicleCategory
-from anyway.widgets.widget import register
 from anyway.widgets.suburban_widgets.sub_urban_widget import SubUrbanWidget
-from anyway.request_params import RequestParams
-import anyway.widgets.widget_utils as widget_utils
+from anyway.widgets.widget import register
+
+TOP_3 = 2
```
Comment (ziv17) [inline anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:18]: I suggest to change the name of this constant to express its usage, rather than its content, like number-of-items or something

File: anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py [hunk 10]
```diff
@@ -28,95 +33,117 @@ def generate_items(self) -> None:
...
+                vehicle_type_dict[VehicleCategory.CAR.value] += count
             elif vehicle_type in VehicleCategory.LARGE.get_codes():
-                driver_types[VehicleCategory.LARGE.value] += count
+                vehicle_type_dict[VehicleCategory.LARGE.value] += count
             elif vehicle_type in VehicleCategory.MOTORCYCLE.get_codes():
-                driver_types[VehicleCategory.MOTORCYCLE.value] += count
+                vehicle_type_dict[VehicleCategory.MOTORCYCLE.value] += count
             elif vehicle_type in VehicleCategory.BICYCLE_AND_SMALL_MOTOR.get_codes():
-                driver_types[VehicleCategory.BICYCLE_AND_SMALL_MOTOR.value] += count
+                vehicle_type_dict[VehicleCategory.BICYCLE_AND_SMALL_MOTOR.value] += count
```
Comment (ziv17) [inline anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:84]: I think electric scooter is not included in any of the categories here. Is it on purpose?

File: anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py [hunk 11]
```diff
@@ -28,95 +33,117 @@ def generate_items(self) -> None:
...
+            sorted(vehicle_type_dict.items(), key=lambda item: item[1], reverse=True)
+        )
+
+        output = {}
+        sum_all_three = 0
+        # Calculate percentage of the top 3
+        for i, (key, value) in enumerate(vehicle_type_dict.items()):
+            output[key] = floor(100 * value / total_count)
+            sum_all_three += output[key]
+            if i >= TOP_3:
```
Comment (ziv17) [inline anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:99]: As the number of returned items is placed in a constant, then for readability I suggest to avoid using the number 3 in comments or variable names.

**Review by ziv17 — APPROVED** [review]
Nice work!

A few comments below

File: anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py (hunk 10 above)
Comment (BarVolunteering) [reply anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:84]: It wasn't part of the requirements, i can add it, @atalyaalon what do you think?

Comment (atalyaalon) [reply anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:84]: @BarVolunteering @ziv17 it's included in the BICYCLE_AND_SMALL_MOTOR category. (see vehicle_type.py)

Comment (BarVolunteering) [reply anyway/widgets/suburban_widgets/accident_count_by_car_type_widget.py:84]: let's discuss in the meeting

**Comment by atalyaalon** [comment]
@BarVolunteering looks good. Merging this one, please create an additional small pr that includes the count distinct fix we discussed
